src/BasicAccess.php

```diff
@@ -53,15 +53,31 @@ public function __construct(
     public function authenticate(ServerRequestInterface $request) : ?UserInterface
     {
         $authHeader = $request->getHeader('Authorization');
-        if (empty($authHeader)) {
+        if (! isset($authHeader[0])) {
```

Comment:
Is `isset` correct or do mean `array_key_exists`?

Reply:
@froschdesign i guess both would work here. I can change it to `array_key_exists` if you like, but why would i want the code to continue here, if it's (somehow) containing `null`?

Reply:
From the PSR-7 spec for `getHeader()`:

> `@return` string[] An array of string values as provided for the given header. If the header does not appear in the message, this method MUST return an empty array.

As such, you will _always_ have an array using this method. The question is if the array is empty; this is why we were using `empty($authHeader)` previously. I see no reason to change that.

Reply:
@weierophinney i agree with that. However, just because it's not-empty strictly does not implies that array index 0 is set. thats why i explicitly check for this, because index 0 is accessed later in the code. If someone manipulated the headers, it could be not empty, but the "first" header is found in index 1 or 2 or whatever.

Do you still want me to revert it to empty?

Reply:
According to [RFC 7235](https://tools.ietf.org/html/rfc7235#section-4.2), the `Authorization` header MUST only be a single header; it cannot be repeated. As such, `$request->getHeader('Authorization')` can only ever return 0 or 1 item in the array.

We can also use `$request->getHeaderLine('Authorization')`, which will return a string with the header contents, or an empty string if the header is not found.

As such, we have two options:

```php
$authHeader = $request->getHeader('Authorization');
if (empty($authHeader)) {
    return null;
}
$value = array_shift($authHeader);
```

**or**

```php
$value = $request->getHeaderLine('Authorization');
if ('' === $value) {
    return null;
```

And from there, use `$value` in the regex when retrieving the credentials from the value.

Reply:
@weierophinney Ah. You would avoid that by using `array_shift()`. I see.

I changed the code to the `getHeaderLine()` suggestion as it's more neat and readable in my opinion. Just pushed it. Thank you.

Reply:
@weierophinney actually, after giving it second thoughts, i changed it back to use `getHeader()` to be easily able to early-exit in case someone sends multiple authorization headers.

After reading [RFC 2617](https://tools.ietf.org/html/rfc2617#section-2), i saw that user-id can be optional as well as the password. making the smallest valid Auth-string only the colon (`:`). I moved some assertions accordingly, as i previously declared some (spec-wise) valid authentication request as invalid.

I also added the test-assertion that invalid data actually never tries to hit the user-repository.

Sorry for the noise. Can you please take a 2nd look now?

If the commit-history is too polluted, i'm happy to squash them if you want.

Thank you very much.

I think this makes for a very robust basic-auth class. :)